src/sources/gcp_pubsub.rs

```diff
@@ -353,15 +454,41 @@ impl PubsubSource {
 
         let (finalizer, mut ack_stream) =
             Finalizer::maybe_new(self.acknowledgements, self.shutdown.clone());
+        let mut pending_acks = 0;
 
         loop {
             tokio::select! {
-                _ = &mut self.shutdown => return State::Shutdown,
-                _ = &mut token_generator.next() => {
+                _ = &mut self.shutdown, if pending_acks == 0 => return State::Shutdown,
+                _ = self.token_generator.changed() => {
                     debug!("New authentication token generated, restarting stream.");
                     break State::RetryNow;
                 },
+                _ = tokio::time::sleep(self.keepalive) => {
+                    if pending_acks == 0 {
+                        // No pending acks, and no new data, so drop
+                        // this stream if we aren't the only active
+                        // one.
+                        if self.concurrency.load(Ordering::Relaxed) > 1 {
+                            info!("Shutting down inactive stream.");
+                            break State::Shutdown;
+                        }
+                        // Otherwise, mark this stream as idle.
+                        busy_flag.store(false, Ordering::Relaxed);
```

Comment:
Could this be boiled down to always break/return?

My thought here is that the logic in the upper select loop will always ensure at least one task is spawned, so this could be used as a more generic way to keep refresh the task and avoid having to do the empty ack trick below to keep the task polling.

Reply:
I wanted to avoid having the outer loop repeatedly reconnect that final connection, hence the logic here. I agree we could simplify the logic by reconnecting instead of sending keepalives, I just don't like it.

Reply:
Yeah, that's fair... was mostly just trying to think through if a clean connection would be a desirable trait, or if the "send us an empty payload with no acks" was something their documentation explicitly recommends for keeping streams alive.

Reply:
I don't recall seeing anything in the docs that mentions either keepalives or timing out connections in streaming pulls.